docs/api/validate.rst

```diff
@@ -0,0 +1,246 @@
+
+    :func:`validate()` is implemented using the stdlib's :func:`functools.singledispatch` decorator, where more specific
+    schemas overload the default implementation with more validation logic.
+
+    ----
+
+    By default, :func:`validate()` compares ``value`` and ``schema`` for equality. This means that simple schema objects
+    like booleans, strings, numbers, None, etc. are validated here, as well as anything unknown.
+
+    Example:
+
+    .. code-block:: python
+
+        schema = validate.Schema(123)
+        assert schema.validate(123) == 123
+        assert schema.validate(123.0) == 123.0
+        schema.validate(456)  # raises ValidationError
+        schema.validate(None)  # raises ValidationError
+
+    :param Any schema: Any kind of object not handled by a more specific validation function
+    :param Any value: The input value
+    :raise ValidationError: If ``value`` and ``schema`` are not equal
+    :return: Unmodified ``value``
+
+.. py:function:: _validate_type(schema, value)
+    :module: streamlink.plugin.api.validate
+
+    :class:`type` validation.
+
+    Checks if ``value`` is an instance of ``schema``.
+
+    Example:
+
+    .. code-block:: python
+
+        schema = validate.Schema(int)
+        assert schema.validate(123) == 123
+        assert schema.validate(True) is True  # `bool` is a subclass of `int`
+        schema.validate("123")  # raises ValidationError
+
+    *This function is included for documentation purposes only! (singledispatch overload)*
+
+    :param type schema: A :class:`type` object
+    :param Any value: The input value
+    :raise ValidationError: If ``value`` is not an instance of ``schema``
+    :return: Unmodified ``value``
+
+.. py:function:: _validate_callable(schema, value)
+    :module: streamlink.plugin.api.validate
+
+    ``Callable`` validation.
+
+    Validates a ``schema`` function where ``value`` gets passed as a single argument.
+
+    Must return a truthy value.
+
+    Example:
+
+    .. code-block:: python
+
+        schema = validate.Schema(
+            lambda val: val < 2,
+        )
+        assert schema.validate(1) == 1
+        schema.validate(2)  # raises ValidationError
+
+    *This function is included for documentation purposes only! (singledispatch overload)*
+
+    :param Callable schema: A function with one argument
+    :param Any value: The input value
+    :raise ValidationError: If ``schema`` returns a non-truthy value
+    :return: Unmodified ``value``
+
+.. py:function:: _validate_sequence(schema, value)
+    :module: streamlink.plugin.api.validate
+
+    :class:`list <builtins.list>`, :class:`tuple`, :class:`set` and :class:`frozenset` validation.
+
+    Each item of ``value`` gets validated against **any** of the items of ``schema``.
+
+    Please note the difference between :class:`list <builtins.list>`
+    and the :class:`ListSchema <_schemas.ListSchema>` validation.
```

Comment:
Potential slight chance of confusion here as people may think you are referring to `streamlink.plugin.api.validate.list` which itself is an alias of `ListSchema` if they aren't familiar with the doc formatting. Not a big deal but it's easier to understand the difference for something like https://github.com/streamlink/streamlink/pull/5652/files#diff-010b6199b0fc1ebe5df5bdb5ee86c9d024fea4dca02b63f388fe11b1d5156599R216 because of the reference to `re.pattern`.

I wish there was a way `<builtins.list>` could easily link to the [python docs](https://docs.python.org/3/library/functions.html#func-list).

Reply:
> people may think you are referring to `streamlink.plugin.api.validate.list`

Maybe, but if they check `validate.list`, they'll see that it's an alias for `ListSchema`, which means that this line compares the same things, which doesn't make much sense and it must be two different things. All members of the `validate` module are always referenced with their module name, never on their own, apart from schema class names.

> I wish there was a way `<builtins.list>` could easily link to the [python docs](https://docs.python.org/3/library/functions.html#func-list).

It's currently disabled. For a good reason, because it's way too noisy with external links everywhere...
https://github.com/streamlink/streamlink/blob/6.3.1/docs/conf.py#L108-L111

Reply:
Works for me.